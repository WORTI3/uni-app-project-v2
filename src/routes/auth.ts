import { Router } from 'express';
import passport from 'passport';
import crypto from "crypto";
import db from "../db";
import { ERROR_MESSAGES, PASSWORD_REGEX } from '../assets/constants';
import { checkValidationResult } from '../middleware/auth';
import { check } from 'express-validator';

const router = Router();

/**
 * GET request handler for the login page.
 * @param {Object} req - The request object (unused).
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function (unused).
 * @returns None
 */
router.get("/login", function (_req, res) {
  res.render("login");
});

/**
 * Handles a POST request to log in a user with a password.
 * @param {string} "/login/password" - the endpoint for the login request
 * @param {function} passport.authenticate - the authentication middleware function
 * @param {object} options - the options object for passport.authenticate
 * @param {string} options.successReturnToOrRedirect - the URL to redirect to upon successful login
 * @param {string} options.failureRedirect - the URL to redirect to upon failed login
 * @param {boolean} options.failureMessage - whether or not to display a failure message
 * @returns None
 */
router.post(
  "/login/password",
  passport.authenticate("local", {
    successReturnToOrRedirect: "/",
    failureRedirect: "/login",
    failureMessage: true,
  })
);

/**
 * Handles a POST request to log out the current user.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {function} next - The next middleware function.
 * @returns None
 */
router.post("/logout", function (req, res, next) {
  req.logout(function (err) {
    if (err) {
      return next(err);
    }
    res.redirect("/");
  });
});

/**
 * GET request handler for the '/signup' route. Renders the 'signup' view.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
router.get("/signup", function (req, res, next) {
  res.render("signup");
});

/**
 * Handles a POST request to sign up a new user. Validates the username and password
 * using the check function from the express-validator library. If the validation passes,
 * the user's password is hashed and salted using the pbkdf2 function from the crypto library.
 * The user's information is then inserted into the database and the user is logged in.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {function} next - The next middleware function.
 * @returns None
 */
router.post(
  "/signup",
  check('username', ERROR_MESSAGES.USERNAME.DEFAULT)
    .isLength({ min: 3 })
    .withMessage(ERROR_MESSAGES.USERNAME.MIN_LENGTH)
    .bail()
    .isLength({ max: 20 })
    .withMessage(ERROR_MESSAGES.PASSWORD.MAX_LENGTH),
  check('password', ERROR_MESSAGES.PASSWORD.DEFAULT)
    .isLength({ min: 5 })
    .withMessage(ERROR_MESSAGES.PASSWORD.MIN_LENGTH)
    .bail()
    .isLength({ max: 20 })
    .withMessage(ERROR_MESSAGES.PASSWORD.MAX_LENGTH)
    .bail()
    .matches(PASSWORD_REGEX)
    .withMessage(ERROR_MESSAGES.PASSWORD.NO_CHARS),
  checkValidationResult, function (req, res, next) {
  const salt = crypto.randomBytes(16);
  crypto.pbkdf2(req.body.password, salt, 310000, 32, "sha256",
    function (err, hashedPassword) {
      if (err) {
        return next(err); // return with db error
      }
      db.run("INSERT INTO users (username, hashed_password, salt) VALUES (?, ?, ?)", [req.body.username, hashedPassword, salt],
        function (err) {
          if (err) {
            return next(err);
          }
          let user = {
            id: this.lastID,
            username: req.body.username,
          };
          req.login(user, function (err) {
            if (err) {
              return next(err);
            }
            res.redirect("/");
          });
        }
      );
    }
  );
});

export default router;
