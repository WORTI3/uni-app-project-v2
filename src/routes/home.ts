import { Request, Response, Router } from 'express';
/**
 * Creates a new instance of an Express router.
 * @returns {Router} - An instance of an Express router.
 */
export const homeRouter = Router();

/**
 * Handles GET requests to the root route.
 * If the user is not logged in, renders the home page.
 * If the user is logged in, renders the index page with the user object.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 */
homeRouter.get('/', (req: Request, res: Response) => {
  // Route handler
  if (!req.user) {
    // Check if user is authenticated
    return res.render('home'); // Render home page
  }

  res.redirect('/dashboard'); // Redirect to dashboard
});
