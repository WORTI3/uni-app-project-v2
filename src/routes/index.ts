import { NextFunction, Request, Response, Router } from 'express';
import {
  ASSET_STATUS,
  ERROR_MESSAGES,
  SUCCESS_MESSAGES,
} from '../assets/constants';
import { check } from 'express-validator';
import { checkValidationResult, ensureAuth, isAdmin } from '../middleware/auth';
import {
  checkAdd,
  checkAll,
  checkEditAdmin,
  checkEditUpdate,
} from '../middleware/routing';
import {
  fetchAssets,
  updateLocalAsset,
  fetchAssetById,
  updateAssetById,
} from '../middleware/asset';
import db from '../db';
import { Session, User } from '../types';

/**
 * Creates a new instance of an Express router.
 * @returns {Router} - An instance of an Express router.
 */
export const indexRouter = Router();

/**
 * GET request handler for the "/all/closed" route.
 * Ensures that the user is logged in and has admin privileges before fetching all assets
 * and filtering out the closed ones. Renders the "index" view with the filtered assets and
 * a flag to indicate that all closed assets should be shown.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
indexRouter.get(
  '/all/closed',
  isAdmin, // checking admin here in case a user navigates via URL
  fetchAssets,
  function (req: Request, res: Response, _next: NextFunction) {
    res.locals.assets = res.locals.assets.filter(function (asset: {
      closed: boolean;
    }) {
      return asset.closed;
    });
    res.render('index', { user: req.user, showAllClosed: true });
  },
);

/**
 * GET request handler for the "/all" route. Ensures that the user is logged in and fetches assets.
 * Filters out any closed assets and renders the "index" view with the user object and showAll flag.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
indexRouter.get(
  '/all',
  isAdmin,
  fetchAssets,
  (req: Request, res: Response, _next: NextFunction) => {
    res.locals.assets = res.locals.assets.filter(function (asset: {
      closed: boolean;
    }) {
      return !asset.closed;
    });
    res.render('index', { user: req.user, showAll: true });
  },
);

indexRouter.get(
  '/dashboard',
  ensureAuth,
  fetchAssets,
  function (req: Request, res: Response, _next: NextFunction) {
    res.render('index', { user: req.user });
  },
);

/**
 * GET request handler for the '/add' route. Renders the 'index' view with the 'addNew' flag set to true.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
indexRouter.get(
  '/add',
  ensureAuth,
  updateLocalAsset,
  (req: Request, res: Response, _next: NextFunction) => {
    return res.render('index', { user: req.user, addNew: true });
  },
);

/**
 * Adds a new asset to the database with the given information.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns {Object} - A redirect to the "/all" page.
 * @throws {Error} - If there is an error inserting the asset into the database.
 */
indexRouter.post(
  '/add',
  ensureAuth,
  check('name', ERROR_MESSAGES.ADD_ISSUE.NAME).isLength({ min: 1 }),
  check('code', ERROR_MESSAGES.ADD_ISSUE.CODE).isLength({ min: 6, max: 6 }),
  check('note', ERROR_MESSAGES.ADD_ISSUE.NOTE).isLength({ min: 3, max: 200 }),
  checkValidationResult,
  (req: Request, res: Response, next: NextFunction) => {
    const user = req.user as User;

    /**
     * Inserts a new asset into the database with the given information.
     */
    const today = new Date().toISOString();
    db.run(
      'INSERT INTO assets (owner_id, owner_name, created, updated, name, code, type, status, note, closed) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)',
      [
        user.id,
        user.username,
        today,
        today,
        req.body.name,
        req.body.code,
        req.body.type,
        ASSET_STATUS.OPEN,
        req.body.note ?? null,
        req.body.closed ? 1 : null,
      ],
      function (err) {
        if (err) {
          console.error('Database errored whilst adding an asset: ', err);
          return next(err);
        }
      },
    );
    const session = req.session as Session;
    session.messages = [SUCCESS_MESSAGES.CREATED];
    session.msgTone = 'positive';
    res.redirect('/all');
  },
);

/**
 * POST request handler for the root route.
 * Ensures that the user is logged in and passes the request through the checkAll, checkAdd middleware.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @returns None
 */
indexRouter.post('/', ensureAuth, checkAll, checkAdd);

/**
 * GET request handler for editing an asset with the given ID.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns {Object} - Renders the index page with the user object and edit flag.
 * @throws {Error} - If the user is not logged in or the asset cannot be found.
 */
indexRouter.get(
  '/:id(\\d+)/edit',
  ensureAuth,
  fetchAssetById,
  updateLocalAsset,
  (req: Request, res: Response, _next: NextFunction) => {
    /**
     * Renders the "index" view with the user and edit parameters.
     * @param {Object} req - The request object.
     * @param {Object} req.user - The user object.
     * @returns The rendered "index" view with the user and edit parameters.
     */
    return res.render('index', { user: req.user, edit: true });
  },
);

/**
 * GET request handler for viewing a specific asset by ID.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns Renders the index page with the user object and readOnly flag set to true.
 * @throws {Error} If the user is not logged in or if there is an error updating or fetching the asset.
 */
indexRouter.get(
  '/:id(\\d+)/view',
  ensureAuth,
  updateAssetById,
  fetchAssetById,
  function (req: Request, res: Response, _next: NextFunction) {
    return res.render('index', { user: req.user, readOnly: true });
  },
);

/**
 * POST request handler for viewing an asset with the given ID.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
indexRouter.post(
  '/:id(\\d+)/view',
  ensureAuth,
  fetchAssetById,
  function (req: Request, res: Response, _next: NextFunction) {
    res.render('index', { user: req.user, readOnly: true });
  },
);

/**
 * Deletes an asset with the given ID and owner ID from the database.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns {Object} - The response object with a redirect to the closed assets page.
 * @throws {Error} - If there is an error deleting the asset from the database.
 */
indexRouter.post(
  '/:id(\\d+)/delete',
  isAdmin,
  /**
   * Deletes an asset from the database with the given ID and owner ID.
   */
  function (req: Request, res: Response, next: NextFunction) {
    const session = req.session as Session;
    db.run(
      'DELETE FROM assets WHERE id = ? AND owner_id = ?',
      [req.params.id, (req.user as User).id],
      function (err) {
        if (err) {
          console.error('Database errored whilst deleting asset: ', err);
          return next(err);
        }
        session.messages = [SUCCESS_MESSAGES.DELETED];
        session.msgTone = 'positive';
        // Redirects the user to the "/all/closed" page with success message.
      },
    );
    res.redirect('/all/closed');
  },
);

/**
 * POST request handler for editing an asset with the given ID.
 *
 * We could validate the type field but not needed for v1.0.0
 *
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
indexRouter.post(
  '/:id(\\d+)/edit',
  ensureAuth,
  check('name', ERROR_MESSAGES.ADD_ISSUE.NAME).isLength({ min: 1 }),
  check('code', ERROR_MESSAGES.ADD_ISSUE.CODE).isLength({ min: 6, max: 6 }),
  check('note', ERROR_MESSAGES.ADD_ISSUE.NOTE).isLength({ min: 3, max: 200 }),
  checkValidationResult,
  checkEditUpdate,
  checkEditAdmin,
  /**
   * Deletes an asset from the database for the authenticated user.
   */
  function (req: Request, res: Response, next: NextFunction) {
    const session = req.session as Session;
    db.run(
      'DELETE FROM assets WHERE id = ? AND owner_id = ?',
      [req.params.id, (req.user as User).id],
      function (err) {
        if (err) {
          console.error('Database errored whilst ammending an asset: ', err);
          return next(err);
        }
        session.messages = [SUCCESS_MESSAGES.DELETED];
        session.msgTone = 'positive';
        // Redirects the user to the "/all" page with success message.
        res.redirect('/all');
      },
    );
  },
);

/**
 * GET request handler for the '/settings' route. Renders the 'settings' view with the
 * authenticated user's information.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
indexRouter.get(
  '/settings',
  ensureAuth,
  function (req: Request, res: Response, _next: NextFunction) {
    res.render('settings', { user: req.user });
  },
);
