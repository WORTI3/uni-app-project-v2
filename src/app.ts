// Dependencies
import express, { ErrorRequestHandler } from 'express';
import nunjucks from 'nunjucks';
import createError from 'http-errors';
import path from 'path';
import cookieParser from 'cookie-parser';
import session from 'express-session';
import csrf from 'csurf';
import sqlite3 from 'sqlite3';
import sqliteStoreFactory from 'express-session-sqlite';
// App routers
import { initPassport } from './middleware/passport';
import { homeRouter } from './routes/home';
import { authRouter } from './routes/auth';
import { indexRouter } from './routes/index';
import { Session } from './types';

const app = express();
// Db
const SqliteStore = sqliteStoreFactory(session);
const store = new SqliteStore({
  driver: sqlite3.Database,
  path: './src/database/sessions.db', // Specify the path to the SQLite database file
  ttl: 86400, // Session Time To Live in seconds
  cleanupInterval: 300000, // Cleanup interval in milliseconds for deleting expired sessions
});

// Nunjucks configuration
nunjucks.configure(['./src/views', './src/_layouts'], {
  autoescape: true,
  express: app,
});
// View engine setup
app.set('view engine', 'njk');
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());

// Static assets (main.css)
app.use(express.static(path.join('./', 'public')));

// Adds CSRF protection to the application.
app.use(csrf({ cookie: true }));

// Init session
app.use(
  session({
    secret: 'nobody knows',
    resave: false,
    saveUninitialized: false,
    store,
  }),
);
// passport middleware
initPassport(app);

/**
 * Sets messages and tone to locals from the current session.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 */
app.use((req, res, next) => {
  const session = req.session as Session;

  if (!session.errorFields) {
    // when no session is present
    const inputValues = {
      username: req.body.username ?? '',
      password: '', // dont want to pass password value through here
    };
    const fields = [];
    for (const [key, value] of Object.entries(inputValues)) {
      fields.push({
        field: key,
        value: value,
      });
    }
    session.errorFields = fields;
  }

  // todo remove messages and use errorFields.error

  if (session.messages || session.errorFields) {
    console.log('error fields:', session.errorFields);

    const msgs = session?.messages ?? [];
    const tone = session.msgTone ?? null;
    res.locals.messages = msgs;
    res.locals.msgTone = tone;
    res.locals.hasMessages =
      Boolean(msgs.length) || Boolean(session.errorFields);
    res.locals.errorFields = session.errorFields;
    session.messages = [];
    // session.errorFields = [];
    session.msgTone = undefined;
  }
  next();
});

/**
 * Middleware function that adds a CSRF token to the response locals object.
 * If the environment is set to "test", a hardcoded token is used instead.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
app.use((req, res, next) => {
  res.locals.csrfToken = req.csrfToken();
  next();
});

// App routes
app.use('/', homeRouter);
app.use('/', indexRouter);
app.use('/', authRouter);

// Catch 404 and forward to error handler
app.use((_req, _res, next) => {
  next(createError(404));
});

// Error handler
const errorHandler: ErrorRequestHandler = (err, req, res, _next) => {
  // Set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};
  res.status(err.status || 500).render('error');
};

app.use(errorHandler);

export default app;
