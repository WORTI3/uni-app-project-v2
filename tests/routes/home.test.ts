import app from '../../src/app';
import request from 'supertest';
import cheerio from 'cheerio';
import { Request, Response } from 'express';
import { homeRouter } from '../../src/routes/home';

const url = '/';

describe('Home Router', () => {
  let req: Partial<Request>;
  let res: Partial<Response>;

  beforeEach(() => {
    req = {};
    res = {
      render: jest.fn(),
      redirect: jest.fn(),
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('GET ' + url, () => {
    it('should show the portal home page when no user present', async () => {
      const result = await request(app).get(url).expect(200);

      expect(result.header['content-type']).toContain('text/html');
      expect(result.text).toMatchSnapshot();
      expect(result.body).toStrictEqual({});
    });

    it('should render the home page when user is not authenticated', async () => {
      const result = await request(app).get(url).expect(200);

      const $ = cheerio.load(result.text);
      const heading = $('h1').text();
      expect(heading).toBe('Welcome to the Faultifier ticket portal');
      const subHeading = $('h2').text();
      expect(subHeading).toBe(
        'to view, create, delete tickets, you must have an account',
      );
      expect($('a').length).toBe(4);
    });
  });

  it('should render home template when user is not logged in', () => {
    // given / when
    homeRouter.stack[0].route.stack[0].handle(
      req as Request,
      res as Response,
      () => {},
    );

    // then
    expect(res.render).toHaveBeenCalledWith('home');
    expect(res.redirect).toHaveBeenCalledTimes(0);
  });

  it('should redirect to dashboard when user is logged in', () => {
    // given
    // Simulate a logged-in user
    req.user = { id: '123', username: 'testuser' };

    // when
    homeRouter.stack[0].route.stack[0].handle(
      req as Request,
      res as Response,
      () => {},
    );

    // then
    expect(res.redirect).toHaveBeenCalledWith('/dashboard');
    expect(res.render).toHaveBeenCalledTimes(0);
  });
});
