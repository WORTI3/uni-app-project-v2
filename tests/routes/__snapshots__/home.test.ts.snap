// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET / should show the portal home page when no user present 1`] = `
"
<!DOCTYPE html>
<html lang="en">
    
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="description" content="Faultifier - a ticket handling admin service.">
        <meta name="theme-color" content="blue">
		<title>Home - Faultifier</title>
		<link rel="stylesheet" type="text/css" href="/main.css">
	</head>

    <div class="flex flex-col h-screen justify-between overflow-hidden">
        <nav class="flex items-center justify-between flex-wrap bg-gray-100 p-6">
    <div class="flex items-center flex-shrink-0 text-picton-blue-600 mr-6">
        <a href="/"><span class="font-bold text-xl tracking-tight font-sans">FAULTIFIER</span></a>
    </div>
    
    
</nav>

        <body class="bg-gradient-to-tr from-picton-blue-600 to-yellow-500 overflow-hidden">
            <main>
<div class="text-center text-white py-20 px-6">
    <h1 class="text-5xl font-bold mt-0 mb-6">Welcome to the Faultifier ticket portal</h1>
    <h2 class="text-md font-semibold mb-8">to view, create, delete tickets, you must have an account</h2>
	<a class="inline-block px-6 py-2.5 bg-picton-blue-800 text-white font-bold text-xs leading-tight uppercase rounded shadow-md hover:bg-picton-blue-700 hover:shadow-lg focus:bg-picton-blue-900 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-picton-blue-900 active:shadow-lg transition duration-150 ease-in-out" data-mdb-ripple="true" data-mdb-ripple-color="light" href="/login" role="button">Sign in</a>
    <p class="inline px-5"> or </p>
    <a class="inline-block px-6 py-2.5 bg-picton-blue-800 text-white font-bold text-xs leading-tight uppercase rounded shadow-md hover:bg-picton-blue-700 hover:shadow-lg focus:bg-picton-blue-900 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-picton-blue-900 active:shadow-lg transition duration-150 ease-in-out" data-mdb-ripple="true" data-mdb-ripple-color="light" href="/signup" role="button">Sign up</a>
</div>
</main>
        </body>

        
        <footer class="lg:text-left bg-gray-100 text-gray-600">
            <div class="text-center p-6">
                <span>Â© 2023 </span>
                <a class="text-gray-600 font-semibold" href="https://github.com/WORTI3">Wortie</a>
            </div>
        </footer>
    </div>
</html>"
`;
